lib/CL/devices/tbb/tbb_scheduler.cc

```diff
@@ -0,0 +1,435 @@
+#include <string.h>
+#include <unistd.h>
+
+#include <tbb/blocked_range3d.h>
+#include <tbb/parallel_for.h>
+#include <tbb/partitioner.h>
+#include <tbb/task_arena.h>
+
+#include "common.h"
+#include "common_utils.h"
+#include "pocl_cl.h"
+#include "pocl_mem_management.h"
+#include "pocl_runtime_config.h"
+#include "pocl_util.h"
+#include "tbb_scheduler.h"
+#include "utlist.h"
+
+static void* pocl_tbb_driver_thread (void *p);
+
+enum partitioner{NONE, AFFINITY, AUTO, SIMPLE, STATIC};
+
+typedef struct scheduler_data_
+{
+  unsigned printf_buf_size;
+  uchar *printf_buf_global_ptr;
+
+  size_t local_mem_size;
+  char *local_mem_global_ptr;
+
+  size_t num_tbb_threads;
+  unsigned grain_size;
+  enum partitioner selected_partitioner;
+
+  _cl_command_node *work_queue __attribute__ ((aligned (HOST_CPU_CACHELINE_SIZE)));
+  POCL_FAST_LOCK_T wq_lock_fast __attribute__ ((aligned (HOST_CPU_CACHELINE_SIZE)));
+
+  pthread_t meta_thread __attribute__ ((aligned (HOST_CPU_CACHELINE_SIZE)));
+  pthread_cond_t wake_meta_thread __attribute__ ((aligned (HOST_CPU_CACHELINE_SIZE)));
+  int meta_thread_shutdown_requested;
+} scheduler_data __attribute__ ((aligned (HOST_CPU_CACHELINE_SIZE)));
+
+static scheduler_data scheduler;
+
+/* External functions declared in tbb_scheduler.h */
+
+void
+tbb_scheduler_init (cl_device_id device)
+{
+  POCL_FAST_INIT (scheduler.wq_lock_fast);
+
+  pthread_cond_init (&(scheduler.wake_meta_thread), NULL);
+
+  scheduler.printf_buf_size = device->printf_buffer_size;
+  assert (device->printf_buffer_size > 0);
+
+  /* safety margin - aligning pointers later (in kernel arg setup)
+   * may require more local memory than actual local mem size.
+   * TODO fix this */
+  scheduler.local_mem_size = device->local_mem_size + device->max_parameter_size * MAX_EXTENDED_ALIGNMENT;
+
+  scheduler.num_tbb_threads = device->max_compute_units;
+
+  /* alloc local memory for all threads making sure the memory for each thread is aligned. */
+  scheduler.printf_buf_size = (1 + (scheduler.printf_buf_size - 1) / MAX_EXTENDED_ALIGNMENT) * MAX_EXTENDED_ALIGNMENT;
+  scheduler.local_mem_size = (1 + (scheduler.local_mem_size - 1) / MAX_EXTENDED_ALIGNMENT) * MAX_EXTENDED_ALIGNMENT;
+  scheduler.printf_buf_global_ptr = reinterpret_cast<uchar*> (pocl_aligned_malloc (MAX_EXTENDED_ALIGNMENT, scheduler.printf_buf_size * scheduler.num_tbb_threads));
+  scheduler.local_mem_global_ptr = reinterpret_cast<char*> (pocl_aligned_malloc (MAX_EXTENDED_ALIGNMENT, scheduler.local_mem_size * scheduler.num_tbb_threads));
+
+  /* create one meta thread to serve as an async interface thread. */
+  pthread_create (&scheduler.meta_thread, NULL, pocl_tbb_driver_thread, NULL);
+
+  scheduler.grain_size = 0;
+  if (pocl_is_option_set("POCL_TBB_GRAIN_SIZE"))
+    {
+      scheduler.grain_size = pocl_get_int_option("POCL_TBB_GRAIN_SIZE", 1);
+      POCL_MSG_PRINT_GENERAL ("TBB: using a grain size of %u\n", scheduler.grain_size);
+    }
+
+  scheduler.selected_partitioner = NONE;
```

Comment:
This would bake TBB's current default policy into PoCL. I would argue against this, because TBB's default policy [has changed in the past](https://github.com/oneapi-src/oneTBB/blob/v2020.3/CHANGES#L2192). Note that TBB's auto_partitioner is an actual partitioner and not a substitute for automatically choosing another partitioner.